pybind11.BUILD

```diff
@@ -46,7 +46,7 @@ cc_library(
     ),
     copts = OPTIONS,
     includes = ["include"],
-    deps = ["@local_config_python//:python_embed"],
+    deps = ["@rules_python//python/cc:current_py_cc_libs"],
```

Comment:
I think the linkopts might be missing? current_py_cc_libs just has the .so files in it, but doesn't set any linkopts itself. Should the rules_python target be setting those linkopts, or is that more the responsibility of the consumer? (this is a domain i'm not well versed in).

I think, worst case, we could determine them at build time. We just need to run an action that runs the interpreter, puts the flags into a file, and then create a CcInfo that adds that file to the command line using `@` inclusion.

Reply:
Looking at `python3-embed.pc` on my workstation, these are the pertinent lines:

```
Libs.private: -ldl 
Libs: -L${libdir} -lpython3.11
Cflags: -I${includedir}/python3.11
```

If `@rules_python//python/cc:current_py_cc_libs` specifies the `.so` files directly, I'm not sure whether these truly matter. I'm inclined to proceed sans `linkopts`, I think, and see whether users of the `pybind11_embed` target report breakage.

Reply:
Thanks for the example flags.

> proceed sans linkopts

Well, it is ultimately up to you, as its your project :). Is there a test that provides some basic assurance?

The -I should be OK -- the underlying libpython target depends on the underlying python_headers target, which sets the includes.

The `-ldl` is because libpython depends on libdl, right? So that probably belongs as part of current_py_cc_libs (why should consumers of python have to specify the extra lib that python needs?)

The `-lpython3.11`: I'm not sure about this one. Since cc_library knows the libs, I would think it'd be smart enough to add `-L` (or otherwise make sure the lib*.so files end up in a findable place). How would it know to add `-llibpython3.11` though?

Reply:
> Well, it is ultimately up to you, as its your project :). Is there a test that provides some basic assurance?

pybind11_bazel currently has no tests, which is... something that we should fix... sooner rather than later, I suppose. I have never used the `pybind11_embed` target myself, however, so I'm unsure what its users' expectations would be.

> The `-ldl` is because libpython depends on libdl, right? So that probably belongs as part of current_py_cc_libs (why should consumers of python have to specify the extra lib that python needs?)

`Libs.private:` is populated [here](https://github.com/python/cpython/blob/5a173efa693a053bf4a059c82c1c06c82a9fa8fb/Misc/python-embed.pc.in#L11) for `python-embed.pc` and [here](https://github.com/python/cpython/blob/5a173efa693a053bf4a059c82c1c06c82a9fa8fb/Misc/python.pc.in#L11) for `python.pc`. Why the latter needs `libdl` is unclear to me, but extracting `LIBS` and specifying them in `linkopts` for the `current_py_cc_libs` target SGTM.

> The `-lpython3.11`: I'm not sure about this one. Since cc_library knows the libs, I would think it'd be smart enough to add `-L` (or otherwise make sure the lib*.so files end up in a findable place). How would it know to add `-llibpython3.11` though?

Is that not implied by depending on the `current_py_cc_libs` target? (Please excuse my display of ignorance here.)

Reply:
> Is that not implied by depending on the current_py_cc_libs target? (Please excuse my display of ignorance here.)

I think that answers my question :) -- rules_python should be setting `-l` in that libs target.

But, just to explain why I asked this at all:

I'm not familiar with this embedding case with shared libraries, so I'm not sure how much current_py_cc_libs is _supposed_ to be telling. Does the use case of "I want the libpython311.so files to be available, but I don't want to tell the linker `-llibpython3.11`" make _any_ sense? That's basically what I'm asking. If that use case is nonsense, then I need to fix rules_python to set `-llibpython3.11`.

If that use case does make sense, then...well I don't know, exactly. I think rules_python would need two `cc_libs` targets (one with `-l` and one without), and then someone else has to be able to be given the supporting pieces to figure out the `3.11` portion of the `-l` string, as well as some way to know which target to use.

Reply:
re: keeping/removing pybind_embed target

Yeah, I'm not sure there's a good reason to keep it long term. In the short term, it probably needs to exist until rules_python sets those `-l` linkopts flags. I'll see if I can get some time this weekend to write a real cc test in rules_python that links and runs and see what happens. All i really need to do is include Python.h and make sure to use a python symbol, right? How hard could that be...

Reply:
> re: keeping/removing pybind_embed target
> 
> Yeah, I'm not sure there's a good reason to keep it long term. In the short term, it probably needs to exist until rules_python sets those `-l` linkopts flags. I'll see if I can get some time this weekend to write a real cc test in rules_python that links and runs and see what happens. All i really need to do is include Python.h and make sure to use a python symbol, right? How hard could that be...

I see there isn't a single `PyObject` in the rules_python repo. :-) (or any C/C++ code)

I'm not sure if that's what you're looking for, but I reduced some simple things I had floating around to a bare minimum:

* https://github.com/rwgk/stuff/tree/master/minimal_embedded_python
* https://github.com/rwgk/stuff/tree/master/minimal_python_extension_module

There are also build_and_run.sh commands that work on current gLinux.

The only dependency is python-dev (Python.h, -lpython3.11).

Reply:
Thanks, @rwgk. I adapted what we talked about in chat into https://github.com/bazelbuild/rules_python/pull/1749

Basically just a cc_test that references some Python symbols, but doesn't actually use them at runtime. This should provide a basic check that the Python C libraries are available. It would be nice to have a binary that could actually run, but properly setting up an embedded binary looks like it might be a bit work.

> it probably needs to exist until rules_python sets those -l linkopts flags.

I think this question in answered now: the cc rules are adding the `-l` flags automatically. Some snippets from the linker args:

```
-rpath
-Xlinker
$ORIGIN/../../../_solib_k8/_U@@_Umain~python~python_U3_U11_Ux86_U64-unknown-linux-gnu_S_S_Clibpython___Ulib
-Xlinker
-rpath
-Xlinker
$ORIGIN/python_libs_linking_test.runfiles/_main/_solib_k8/_U@@_Umain~python~python_U3_U11_Ux86_U64-unknown-linux-gnu_S_S_Clibpython___Ulib
-Lbazel-out/k8-fastbuild/bin/_solib_k8/_U@@_Umain~python~python_U3_U11_Ux86_U64-unknown-linux-gnu_S_S_Clibpython___Ulib
...
-lpython3.11
-l:libpython3.11.so.1.0
...
```

The cc_library target that rules_python is defining doesn't explicitly say any of that, so the cc rules are adding that themselves. I don't see e.g. `-ldl`, but that's a constant name instead of a X.Y name that varies depending on the python version; worst case, users can add it themselves to their binary or intermediary cc_library targets.

Reply:
https://github.com/pybind/pybind11_bazel/issues/71#issuecomment-1939537544 captures my intention to do away with this. This target, I mean. :)